python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1304,6 +1305,30 @@ def update_replace_dict_and_conflicts(
         merged_attrs = merge_dicts(self.get_attributes_dict(), replaced_attributes)
         return self.__class__.dagster_internal_init(**merged_attrs)
 
+    def map_asset_specs(self, fn: Callable[[AssetSpec], AssetSpec]) -> "AssetsDefinition":
+        mapped_specs = []
+        for spec in self.specs:
+            mapped_spec = fn(spec)
+            if mapped_spec.key != spec.key:
+                raise DagsterInvalidDefinitionError(
+                    f"Asset key {spec.key.to_user_string()} was changed to "
+                    f"{mapped_spec.key.to_user_string()}. Mapping function must not change keys."
+                )
+            if (
+                # check reference equality first for performance
```

Comment:
ran a little benchmark to see if short circuiting list equality for reference equality actually helps:
```python
from datetime import datetime
from random import random


def random_list(n):
    return [random() for _ in range(n)]


random_lists = [random_list(1000) for _ in range(10000)]

start = datetime.now()
for _ in range(1000):
    for lst in random_lists:
        assert lst is lst or lst == lst


end = datetime.now()
print(end.timestamp() - start.timestamp())
```

short-circuiting with reference equality checks resulted in 10x improvment